packages/@aws-cdk/aws-eks/lib/service-account.ts

```diff
@@ -0,0 +1,74 @@
+import { FederatedPrincipal, IRole, OpenIdConnectProvider, Role } from '@aws-cdk/aws-iam';
+import { Construct } from '@aws-cdk/core';
+import { Cluster } from './cluster';
+
+/**
+ * Service Account
+ */
+export interface ServiceAccountOptions {
+  /**
+   * The cluster to apply the patch to.
+   */
+  readonly name: string;
+  /**
+   * The cluster to apply the patch to.
+   */
+  readonly namespace: string;
+}
+
+/**
+ * Service Account
+ */
+export interface ServiceAccountProps extends ServiceAccountOptions {
+  /**
+   * The cluster to apply the patch to.
+   * [disable-awslint:ref-via-interface]
+   */
+  readonly cluster: Cluster;
+}
+
+/**
+ * Service Account
+ */
+export class ServiceAccount extends Construct {
+
+  /**
+   * The role the service account is linked to.
+   */
+  public readonly role: IRole;
+
+  constructor(scope: Construct, id: string, props: ServiceAccountProps) {
+    super(scope, id);
+
+    const { cluster, name, namespace } = props;
+
+    let provider = cluster.node.tryFindChild('OpenIdConnectProvider') as OpenIdConnectProvider;
+    if (!provider) {
+      if (!cluster.kubectlEnabled) {
+        throw new Error('Cannot specify a OpenID Connect Provider if kubectl is disabled');
+      }
+      provider = new OpenIdConnectProvider(cluster, 'OpenIdConnectProvider', {
+        url: props.cluster.clusterOpenIdConnectIssuerUrl!,
+      });
+    }
+
+    this.role = new Role(this, 'Role', {
+      assumedBy: new FederatedPrincipal(provider.openIdConnectProviderArn, {}, 'sts:AssumeRoleWithWebIdentity'),
```

Comment:
Should we add an `iam.WebIdentityPrincipal` that will wrap this? Alternatively, we can me `OpenIdConnectProvider` an `iam.IPrincipal` or add an `openIdConnectProvider.principal`

Reply:
It's actually a common pattern and usage with IAM. Only the condition would vary but it's always a FederatedPrincipal with action sts:AssumeRoleWithWebIdentity.

Might it be a solutions as well to add the conditions using a custom resource to overcome the lack of support for intrinsic functions as map keys? The conditions are important as security feature. Without it allows every pod to assume the role. The condition pervents this.

Reply:
Yes, definitely sounds like heavy lifting we can offer. Do you want to take a whack at it?

Reply:
@eladb I'm busy with adding the WebIdentityPrincial which make is a little bit easier to work with. However I'm still struggling a bit the dynamic map key of the EKS version of the WebIdentityPrincipal. I can't pass the value directly from the cluster as it is an attribute reference instead of the real value. This means I have two options to pursuit:

1. Create a dedicated custom resource to change the condition based on what is given by the ServiceAccount construct in either the iam module or eks.
2. Create a policy document template custom resource in iam which allows you to replace `${Foo}` with `Bar` in a policy document. 

What I noticed in the documentation that there are 4 common ones (Cognito, Amazon, Facebook & Google) which we can create a class off and add the conditions using properties. But this is something for a separate PR.